iris-mpc-cpu/src/hawkers/aby3/aby3_store.rs

```diff
@@ -483,6 +495,150 @@ impl VectorStore for Aby3Store {
+            .await?;
+        compaction_list
+            .first()
+            .ok_or_eyre("Unexpected: no compacted neighborhood returned from batch processing")
+            .cloned()
+    }
+
+    async fn compact_neighborhood_batch(
+        &mut self,
+        base_nodes: &[Self::VectorRef],
+        neighborhoods: &[Vec<Self::VectorRef>],
+        max_sizes: &[usize],
+    ) -> Result<Vec<Vec<Self::VectorRef>>> {
+        if base_nodes.len() != neighborhoods.len() || base_nodes.len() != max_sizes.len() {
+            bail!("Lists of base nodes, neighborhoods, and max sizes must have equal sizes");
+        }
+
+        let base_node_queries = self.vectors_as_queries(base_nodes.to_vec()).await;
+        let query_vec_pairs = izip!(base_node_queries, neighborhoods.iter())
+            .flat_map(|(q, nbhd)| nbhd.iter().map(move |nb| (q.clone(), *nb)))
+            .collect_vec();
+        let distances = self.eval_distance_pairs(&query_vec_pairs).await?;
+        let id_distances = neighborhoods
+            .iter()
+            .flatten()
+            .zip(distances)
+            .map(|(vector_id, distance)| (vector_id.serial_id(), distance))
+            .collect_vec();
+        let id_versions: BTreeMap<_, _> = neighborhoods
+            .iter()
+            .enumerate()
+            .flat_map(|(idx, nbhd)| {
+                nbhd.iter()
+                    .map(move |vector_id| ((idx, vector_id.serial_id()), vector_id.version_id()))
+            })
+            .collect();
+
+        // Construct aggregated selection networks for top-k selection over all neighborhoods
+        let mut total_items: usize = 0;
+        let mut batched_network = SwapNetwork::new();
+        for (nbhd, target_size) in izip!(neighborhoods.iter(), max_sizes.iter()) {
+            let current_size = nbhd.len();
+
+            // Constructed network is already optimized for the case of k > n - k
+            let mut network = min_k_batcher_sort_network(current_size, *target_size)?;
+
+            // Merge individual swap network into overall batch network
+            let network_shift_amount = isize::try_from(total_items)?;
+            network.shift(network_shift_amount)?;
+            batched_network = SwapNetwork::merge_parallel(batched_network, network);
+
+            total_items += current_size;
+        }
+
+        // Oblivious application of batched selection networks
+        let res_id_distances =
+            apply_oblivious_swap_network(self, &id_distances, &batched_network).await?;
+
+        // Truncate results and unpack into individual vectors
+        let mut unshuffled_truncated_shares = Vec::with_capacity(neighborhoods.len());
+        let mut base_idx = 0;
+        for (nbhd, max_size) in izip!(neighborhoods.iter(), max_sizes.iter()) {
+            let n_keep = usize::min(nbhd.len(), *max_size);
+            let nbhd_shares = res_id_distances[base_idx..base_idx + n_keep].to_vec();
```

Comment:
I feel there's generally a lot of flatenning and unflatenning in batched implementations. Wonder if there's a better framework for it. Hopefully these allocations don't show up in the flamegraphs. 

(No action needed)